File: python/iforevents/flatten.py
"""Nested dict flattening with ``_``, as the Flutter package does."""
from __future__ import annotations

from typing import Any, Dict


def flatten(data: Dict[str, Any], prefix: str = "") -> Dict[str, Any]:
    """``{"a": {"b": 1}}`` becomes ``{"a_b": 1}``; lists and other values are kept."""
    out: Dict[str, Any] = {}
    for key, value in data.items():
        name = f"{prefix}_{key}" if prefix else str(key)
        if isinstance(value, dict):
            out.update(flatten(value, name))
        else:
            out[name] = value
    return out

File: js/packages/core/src/flatten.ts
import type { Properties } from "./types";

/**
 * Flattens nested plain objects with `_`, as the Flutter package does:
 * `{ a: { b: 1 } }` becomes `{ a_b: 1 }`. Arrays, dates and class instances
 * are kept as values.
 */
export function flatten(input: Properties, prefix = ""): Properties {
  const out: Properties = {};
  for (const [key, value] of Object.entries(input)) {
    const name = prefix ? `${prefix}_${key}` : key;
    if (isPlainObject(value)) Object.assign(out, flatten(value as Properties, name));
    else out[name] = value;
  }
  return out;
}

export function isPlainObject(value: unknown): value is Properties {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}
